File: research/src/game2048/board.py
"""盤面ユーティリティ (SPEC.md #10, #12)。web/src/game/board.ts に対応。"""

from __future__ import annotations

Board = list[int]

BOARD_SIZE = 4
CELL_COUNT = BOARD_SIZE * BOARD_SIZE


def create_empty_board() -> Board:
    return [0] * CELL_COUNT


def clone_board(board: Board) -> Board:
    return board.copy()


def get_empty_cells(board: Board) -> list[int]:
    return [i for i, value in enumerate(board) if value == 0]


def get_max_tile(board: Board) -> int:
    return max(board) if board else 0


def serialize_board(board: Board) -> str:
    return ",".join(str(v) for v in board)


def boards_equal(a: Board, b: Board) -> bool:
    return a == b

File: web/src/game/board.ts
import type { Board } from "./types";

/** 盤面サイズは可変 (issue #16, #17)。既定は従来通りの 4x4。 */
export const DEFAULT_BOARD_SIZE = 4;
export const MIN_BOARD_SIZE = 3;
export const MAX_BOARD_SIZE = 5;

/** 一次元配列の長さから一辺のマス数を求める */
export function boardSizeOf(board: Board): number {
  return Math.round(Math.sqrt(board.length));
}

export function createEmptyBoard(size: number = DEFAULT_BOARD_SIZE): Board {
  return new Array(size * size).fill(0);
}

export function cloneBoard(board: Board): Board {
  return board.slice();
}

/** 空きマスのインデックス一覧を返す (SPEC.md #10.1) */
export function getEmptyCells(board: Board): number[] {
  const cells: number[] = [];
  for (let i = 0; i < board.length; i++) {
    if (board[i] === 0) cells.push(i);
  }
  return cells;
}

/** 盤面中の最大タイル値を返す (SPEC.md #10.1) */
export function getMaxTile(board: Board): number {
  let max = 0;
  for (const value of board) {
    if (value > max) max = value;
  }
  return max;
}

/** キャッシュキー生成等に使う文字列表現 (SPEC.md #11.8) */
export function serializeBoard(board: Board): string {
  return board.join(",");
}

export function boardsEqual(a: Board, b: Board): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}
